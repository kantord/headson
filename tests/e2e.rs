use std::fs;
use std::path::Path;

use assert_cmd::Command;
use insta::assert_snapshot;

fn run_case(path: &Path, template: &str, n: u32) -> String {
    let input = fs::read_to_string(path).expect("read fixture");
    let mut cmd = Command::cargo_bin("headson").expect("bin");
    let output = cmd
        .arg("-n")
        .arg(n.to_string())
        .arg("-f")
        .arg(template)
        .write_stdin(input)
        .output()
        .expect("run");
    String::from_utf8_lossy(&output.stdout).into_owned()
}

#[test]
fn e2e_parametric() {
    let dir = Path::new("tests/e2e_inputs");
    let templates = ["json", "pseudo", "js"];
    let budgets = [10u32, 100u32, 250u32, 1000u32, 10000u32];
    for entry in fs::read_dir(dir).expect("list dir") {
        let entry = entry.unwrap();
        if entry.file_type().unwrap().is_file() {
            let name = entry.file_name().into_string().unwrap();
            for &n in &budgets {
                for tmpl in templates {
                    let stdout = run_case(&entry.path(), tmpl, n);
                    assert_snapshot!(
                        format!(
                            "e2e_{}__{}__n{}",
                            name.replace('.', "_"),
                            tmpl,
                            n
                        ),
                        stdout
                    );
                }
            }
        }
    }
}
