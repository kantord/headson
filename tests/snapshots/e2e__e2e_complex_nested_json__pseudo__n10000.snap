---
source: tests/e2e.rs
expression: stdout
---
{
  "users": [
    {
      "name": "Alice Wonderland",
      "profile": {
        "bio": "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.",
        "social": {
          "twitter": "@alice",
          "github": "alice",
          "web": "https://example.com/users/alice"
        }
      },
      "id": 1,
      "roles": [
        "admin",
        "editor",
        "viewer"
      ]
    },
    {
      "profile": {
        "bio": "Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
        "social": {
          "github": "bobbuilder",
          "twitter": "@buildWithBob",
          "web": "https://bob.example.org"
        }
      },
      "roles": [
        "contributor",
        "viewer"
      ],
      "name": "Bob Builder",
      "id": 2
    }
  ],
  "meta": {
    "next": null,
    "tags": [
      "alpha",
      "beta",
      "gamma",
      "delta",
      "epsilon"
    ],
    "page": 1,
    "pageSize": 20
  }
}
